drivers/pwm/pwm_max31790.c

```diff
@@ -0,0 +1,685 @@
+	MAX37190_TACH12COUNTMSB = 0x2E,
+	MAX37190_TACH12COUNTLSB = 0x2F,
+	MAX37190_PWMOUT1DUTYCYCLEMSB = 0x30,
+	MAX37190_PWMOUT1DUTYCYCLELSB = 0x31,
+	MAX37190_PWMOUT2DUTYCYCLEMSB = 0x32,
+	MAX37190_PWMOUT2DUTYCYCLELSB = 0x33,
+	MAX37190_PWMOUT3DUTYCYCLEMSB = 0x34,
+	MAX37190_PWMOUT3DUTYCYCLELSB = 0x35,
+	MAX37190_PWMOUT4DUTYCYCLEMSB = 0x36,
+	MAX37190_PWMOUT4DUTYCYCLELSB = 0x37,
+	MAX37190_PWMOUT5DUTYCYCLEMSB = 0x38,
+	MAX37190_PWMOUT5DUTYCYCLELSB = 0x39,
+	MAX37190_PWMOUT6DUTYCYCLEMSB = 0x3A,
+	MAX37190_PWMOUT6DUTYCYCLELSB = 0x3B,
+	MAX37190_PWMOUT1TARGETDUTYCYCLEMSB = 0x40,
+	MAX37190_PWMOUT1TARGETDUTYCYCLELSB = 0x41,
+	MAX37190_PWMOUT2TARGETDUTYCYCLEMSB = 0x42,
+	MAX37190_PWMOUT2TARGETDUTYCYCLELSB = 0x43,
+	MAX37190_PWMOUT3TARGETDUTYCYCLEMSB = 0x44,
+	MAX37190_PWMOUT3TARGETDUTYCYCLELSB = 0x45,
+	MAX37190_PWMOUT4TARGETDUTYCYCLEMSB = 0x46,
+	MAX37190_PWMOUT4TARGETDUTYCYCLELSB = 0x47,
+	MAX37190_PWMOUT5TARGETDUTYCYCLEMSB = 0x48,
+	MAX37190_PWMOUT5TARGETDUTYCYCLELSB = 0x49,
+	MAX37190_PWMOUT6TARGETDUTYCYCLEMSB = 0x4A,
+	MAX37190_PWMOUT6TARGETDUTYCYCLELSB = 0x4B,
+	MAX37190_USERBYTE6 = 0x4C,
+	MAX37190_USERBYTE7 = 0x4D,
+	MAX37190_USERBYTE8 = 0x4E,
+	MAX37190_USERBYTE9 = 0x4F,
+	MAX37190_TACH1TARGETCOUNTMSB = 0x50,
+	MAX37190_TACH1TARGETCOUNTLSB = 0x51,
+	MAX37190_TACH2TARGETCOUNTMSB = 0x52,
+	MAX37190_TACH2TARGETCOUNTLSB = 0x53,
+	MAX37190_TACH3TARGETCOUNTMSB = 0x54,
+	MAX37190_TACH3TARGETCOUNTLSB = 0x55,
+	MAX37190_TACH4TARGETCOUNTMSB = 0x56,
+	MAX37190_TACH4TARGETCOUNTLSB = 0x57,
+	MAX37190_TACH5TARGETCOUNTMSB = 0x58,
+	MAX37190_TACH5TARGETCOUNTLSB = 0x59,
+	MAX37190_TACH6TARGETCOUNTMSB = 0x5A,
+	MAX37190_TACH6TARGETCOUNTLSB = 0x5B,
+	MAX37190_USERBYTE10 = 0x5C,
+	MAX37190_USERBYTE11 = 0x5D,
+	MAX37190_USERBYTE12 = 0x5E,
+	MAX37190_USERBYTE13 = 0x5F,
+	MAX37190_WINDOW1 = 0x60,
+	MAX37190_WINDOW2 = 0x61,
+	MAX37190_WINDOW3 = 0x62,
+	MAX37190_WINDOW4 = 0x63,
+	MAX37190_WINDOW5 = 0x64,
+	MAX37190_WINDOW6 = 0x65,
+	MAX37190_USERBYTE14 = 0x66,
+	MAX37190_USERBYTE15 = 0x67,
+};
+
+#define MAX37190_GET_VALUE(value, pos, length) FIELD_GET(GENMASK(pos + length - 1, pos), value)
+#define MAX37190_SET_VALUE(target, value, pos, length)                                             \
+	target &= ~GENMASK(pos + length - 1, pos);                                                 \
+	target |= FIELD_PREP(GENMASK(pos + length - 1, pos), value)
+
+#define MAX37190_GLOBALCONFIGURATION_STANDBY_LENGTH 1
+#define MAX37190_GLOBALCONFIGURATION_STANDBY_POS    7
+#define MAX37190_GLOBALCONFIGURATION_STANDBY_GET(value)                                            \
+	MAX37190_GET_VALUE(value, MAX37190_GLOBALCONFIGURATION_STANDBY_POS,                        \
+			   MAX37190_GLOBALCONFIGURATION_STANDBY_LENGTH)
+#define MAX37190_GLOBALCONFIGURATION_STANDBY_SET(target, value)                                    \
+	MAX37190_SET_VALUE(target, value, MAX37190_GLOBALCONFIGURATION_STANDBY_POS,                \
+			   MAX37190_GLOBALCONFIGURATION_STANDBY_LENGTH)
```

Comment:
This seems like a very odd way of implementing register access and changes with the Zephyr API, would you be able to use functions instead of macros and make use of `i2c_reg_update_byte_dt` for bitmasks?

Reply:
I prefer to have only one write access per register, hence the possibility to set all valid bits of a register byte before it is written out. Reduces the amount of I2C transactions which are required to communicate with the IC.

Regarding the macros vs. functions, I also think using functions would be better readable. But on a rather resource constrained system like an MCU I consider macros being the preferrable aproach.

Reply:
I know for a fact that having multiple operations per register set will at least have a measurable negative impact on my use case. Therefore I would definitely try to write all values for one register at once anyway. If you prefer functions I can replace the macros like MAX37190_GLOBALCONFIGURATION_STANDBY_GET/SET with functions, but I would still keep MAX37190_GLOBALCONFIGURATION_STANDBY_LENGTH and MAX37190_GLOBALCONFIGURATION_STANDBY_POS to avoid the usage of magic numbers for bit positions and meanings. Then I doubt that the result will be more readable, comparing

`max31790_set_globalconfiguration_standby(register_value, 0b1);`

with

`MAX37190_GLOBALCONFIGURATION_STANDBY_SET(register_value, 0b1);`.

Regarding the unused macros, I went through all by this driver used registers and defined these macros for all the definitions contained in them. Therefore, some registers are still not listed, as they are not used at all by the driver. But for the sake of consistency and completeness I added the unused register-definitions for the registers which are used at least once. If you prefer not to have them here I can also get rid of them again.

Reply:
`max31790_set_globalconfiguration_standby(register_value, 0b1);` should just be `global_configuration |= BIT(MAX37190_GLOBALCONFIGURATION_STANDBY_POS)`.

Reply:
Bitwise OR-ing would work fine for single bit configuration, but a bunch of these configurations spread over multiple bits. For these I think it is more clear to have one macro call to set them, instead of OR-ing into a byte multiple separate bits. Therefore I'm gonna need some macros anyway, and then I prefer to handle all these register accesses similarly to have some consistency.

Reply:
From what I can tell there's only a few multi-bit ones that are used. Look I'm sure this makes perfect sense if you are used to work with this pattern on all your drivers, but that's not how most drivers are written here. The Linux driver from the same device does everything with just a handful of very compact register definitions and a couple bitmasks (https://lxr.missinglinkelectronics.com/linux/drivers/hwmon/max31790.c#L17) and Zephyr drivers tend to follow a similar pattern. I understand that there's an aspect of personal preference but we need some degree of uniformity in the code base for the sake of maintainability. This block of macro is almost a micro-hal within the driver, it's unnecessarily verbose and many of those are not used or tested anyway.

Reply:
I agree with @fabiobaltieri on this one. The current approach does not match the coding style for for doing this type of register value modifications as used in the rest of the project. It comes across as being too verbose.

Reply:
Okay, will change it.

Reply:
done